src/Compilers/CSharp/Portable/BoundTree/DecisionTreeBuilder.cs

```diff
@@ -26,17 +26,63 @@ internal abstract class DecisionTreeBuilder
         protected readonly Symbol _enclosingSymbol;
         protected readonly Conversions _conversions;
         protected HashSet<DiagnosticInfo> _useSiteDiagnostics = new HashSet<DiagnosticInfo>();
+        private Dictionary<TypeSymbol, LocalSymbol> localByType = new Dictionary<TypeSymbol, LocalSymbol>();
 
         protected DecisionTreeBuilder(
             Symbol enclosingSymbol,
+            SyntaxNode syntax,
             Conversions conversions)
         {
             this._enclosingSymbol = enclosingSymbol;
+            this.Syntax = syntax;
             this._conversions = conversions;
         }
 
         protected SyntaxNode Syntax { private get; set; }
 
+        private LocalSymbol PatternMatchingTemp(TypeSymbol type)
+        {
+            LocalSymbol temp;
+            if (!localByType.TryGetValue(type, out temp))
+            {
+                temp = new SynthesizedLocal(_enclosingSymbol as MethodSymbol, type, SynthesizedLocalKind.PatternMatching, Syntax);
```

Comment:
Yes, the parameter is "optional" in the constructor.

While it would be possible to change this to a cast now, that would be an error when we introduce a match expression, because in that case the enclosing symbol could be a field (for a field initializer). I'd rather leave this as is so it is future-proof.